single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,751 @@
+            == secret_label
+        ):
+            scope = Scope.APP
+        elif (
+            generate_secret_label(self.charm.app.name, self.charm.peer_rel_name, Scope.UNIT)
+            == secret_label
+        ):
+            scope = Scope.UNIT
+        else:
+            logging.debug("Secret %s changed, but it's unknown", secret_id)
+            return
+        logging.debug("Secret %s for scope %s changed, refreshing", secret_id, scope)
+        self.state.secrets.get(scope)
+
+        # Always update the PBM and mongodb exporter configuration so that if
+        # the secret changed, the configuration is updated and will still work
+        # afterwards.
+        self.mongodb_exporter_config_manager.connect()
+        self.backup_manager.connect()
+
+    @override
+    def on_relation_departed(self, departing_unit: Unit | None) -> None:
+        """Handles the relation departed events."""
+        if not self.charm.unit.is_leader() or departing_unit == self.charm.unit:
+            return
+        if self.state.upgrade_in_progress:
+            # do not defer or return here, if a user removes a unit, the config will be incorrect
+            # and lead to MongoDB reporting that the replica set is unhealthy, we should make an
+            # attempt to fix the replica set configuration even if an upgrade is occurring.
+            logger.warning(
+                "Removing replicas during an upgrade is not supported. The charm may be in a broken, unrecoverable state"
+            )
+        self.update_hosts()
+
+    @override
+    def on_storage_attached(self) -> None:  # pragma: nocover
+        """Handler for `storage_attached` event.
+
+        This should handle fixing the permissions for the data dir.
+        """
+        if self.substrate == Substrates.VM:
+            self.workload.exec(["chmod", "-R", "770", f"{self.workload.paths.common_path}"])
+            self.workload.exec(
+                [
+                    "chown",
+                    "-R",
+                    f"{self.workload.users.user}:{self.workload.users.group}",
+                    f"{self.workload.paths.common_path}",
+                ]
+            )
+
+    @override
+    def on_storage_detaching(self) -> None:
+        """Before storage detaches, allow removing unit to remove itself from the set.
+
+        If the removing unit is primary also allow it to step down and elect another unit as
+        primary while it still has access to its storage.
+        """
+        if self.state.upgrade_in_progress:
+            # We cannot defer and prevent a user from removing a unit, log a warning instead.
+            logger.warning(
+                "Removing replicas during an upgrade is not supported. The charm may be in a broken, unrecoverable state"
+            )
+        # A single replica cannot step down as primary and we cannot reconfigure the replica set to
+        # have 0 members.
```

Comment:
And this is something which is copy pasted from former code: [see here](https://github.com/canonical/mongodb-operator/blob/6/edge/src/charm.py#L610-L611)
This part is implemented in another PR (because it needs sharding).
As in other places, I tried as much as possible not to change behaviors in the refactor/rewrite to limit the risks of breaking/failing and putting the project at risk.